lib/collection/src/grouping/aggregator.rs

```diff
@@ -0,0 +1,466 @@
+        let payload_values: Vec<_> = point
+            .payload
+            .as_ref()
+            .map(|p| {
+                p.get_value(&self.grouped_by)
+                    .values()
+                    .into_iter()
+                    .flat_map(|v| match v {
+                        Value::Array(arr) => arr.iter().collect(),
+                        _ => vec![v],
+                    })
+                    .collect()
+            })
+            .ok_or(KeyNotFound)?;
+
+        let group_keys = payload_values
+            .into_iter()
+            .map(GroupId::try_from)
+            .collect::<Result<Vec<GroupId>, ()>>()
+            .map_err(|_| BadKeyType)?;
+
+        let unique_group_keys: Vec<_> = group_keys.into_iter().unique().collect();
+
+        for group_key in unique_group_keys {
+            let group = self
+                .groups
+                .entry(group_key.clone())
+                .or_insert_with(|| HashMap::with_capacity(self.max_group_size));
+
+            let entry = group.entry(point.id);
+
+            // if the point is already in the group, check if it has newer version
+            match entry {
+                Entry::Occupied(mut o) => {
+                    if o.get().version < point.version {
+                        o.insert(point.clone());
+                    }
+                }
+                Entry::Vacant(v) => {
+                    v.insert(point.clone());
+                }
+            }
+
+            if group.len() == self.max_group_size {
+                self.full_groups.insert(group_key.clone());
+            }
+
+            // Insert score if
+            self.group_max_scores
+                .entry(group_key.clone())
+                .and_modify(|e| *e = point.score.max(*e))
+                .or_insert(point.score);
+
+            self.group_min_scores
+                .entry(group_key.clone())
+                .and_modify(|e| *e = point.score.min(*e))
+                .or_insert(point.score);
+        }
+        Ok(())
+    }
+
+    /// Adds multiple points to the group that they corresponds based on the group_by field, assumes that the points always have the grouped_by field, else it just ignores them
+    pub(super) fn add_points(&mut self, points: &[ScoredPoint]) {
+        for point in points {
+            match self.add_point(point.to_owned()) {
+                Ok(()) | Err(KeyNotFound | BadKeyType) => continue, // ignore points that don't have the group_by field
+            }
+        }
+    }
+
+    #[cfg(test)]
+    pub(super) fn len(&self) -> usize {
+        self.groups.len()
+    }
+
+    /// Return `max_groups` number of keys of the groups with the best score
+    fn best_group_keys(&self) -> Vec<&GroupId> {
+        match self.order {
+            Order::LargeBetter => {
+                self.group_max_scores
+                    .iter()
+                    .sorted_by_key(|(_, score)| OrderedFloat(**score))
```

Comment:
we could sort by the opposite comparator right away to avoid having to reverse the iterator